alf/environments/wrappers.py

```diff
@@ -27,44 +29,144 @@
+                            high = np.transpose(
+                                np.concatenate(
+                                    [np.transpose(sp.high)] * stack_size))
+                        stacked_space = gym.spaces.Box(
+                            low=np.array(low),
+                            high=np.array(high),
+                            dtype=sp.dtype)
+
+                    elif name == 'sentence':
+                        assert isinstance(sp, gym.spaces.MultiDiscrete)
+                        nvec = [
+                            stack_size * sp.nvec[0],
+                        ] + list(sp.nvec[1:])
+                        stacked_space = gym.spaces.MultiDiscrete(nvec)
+
+                    else:
+                        assert name not in self._white_list_fields, (
+                            'Please' + ' update _white_list_fields.')
+                        logging.warning(
+                            'Stacking of {} field' +
+                            ' is not supported.  Not stacking.'.format(name))
+                        stacked_space = sp
+                    self.observation_space.spaces[name] = stacked_space
+
+                else:  # name not in fields_to_stack:
+                    self.observation_space.spaces[name] = sp
+
+        else:  # raw_space is a simple gym.spaces.Box, not a dict
+            assert isinstance(raw_space, gym.spaces.Box)
+            sp = raw_space
+            low = np.concatenate([sp.low] * stack_size)
+            high = np.concatenate([sp.high] * stack_size)
+            # Assuming we are processing image input
+            assert channel_order in ['channels_last', 'channels_first']
+            if channel_order == 'channels_last':
+                low = np.transpose(
+                    np.concatenate([np.transpose(sp.low)] * stack_size))
+                high = np.transpose(
+                    np.concatenate([np.transpose(sp.high)] * stack_size))
+            stacked_space = gym.spaces.Box(
+                low=np.array(low), high=np.array(high), dtype=sp.dtype)
+            self.observation_space = stacked_space
 
     def __getattr__(self, name):
         """Forward all other calls to the base environment."""
         return getattr(self.env, name)
 
     def _generate_observation(self):
-        if self._channel_order == 'channels_last':
-            return np.concatenate(self._frames, axis=-1)
-        else:
-            return np.concatenate(self._frames, axis=0)
+        if self._fields_to_stack:
+            result = collections.OrderedDict()
+            for frame in self._frames:
+                for key, field in frame.items():
+                    if key not in result:
+                        result[key] = collections.deque(
+                            maxlen=self._stack_size)
+                    result[key].append(field)
+            for key in result:
+                if key in self._fields_to_stack:
+                    if (key == 'image'
```

Comment:
This can be a problem if the key is not "image" but still a image. It's probably better to always respect _channel order